File: agent_service/schemas/unified_search.py
"""四库联合搜索请求与响应 DTO。

使用说明：REST、gRPC 和前端共享本模块定义的来源常量与输出结构；业务搜索由
``UnifiedSearchService`` 执行，DTO 不包含检索逻辑。
"""

from __future__ import annotations

from typing import Any, Literal

from sqlmodel import Field, SQLModel

from agent_service.core.agent_config import DEFAULT_BUSINESS_LIMITS

UnifiedSearchSource = Literal["files", "library", "components", "literature"]
UnifiedSearchMode = Literal["title", "fulltext", "semantic"]


class UnifiedSearchRequest(SQLModel):
    """描述一次四库联合搜索请求。"""

    user_id: str = Field(min_length=DEFAULT_BUSINESS_LIMITS.nonempty_min_length)
    query: str = Field(min_length=DEFAULT_BUSINESS_LIMITS.nonempty_min_length)
    sources: list[UnifiedSearchSource]
    fulltext: bool = True
    semantic: bool = False


class UnifiedSearchResultOut(SQLModel):
    """统一列表和分裂原生卡片共同消费的一条结果。"""

    id: str
    source: UnifiedSearchSource
    title: str
    snippet: str = ""
    locator: str = ""
    updated_at: str = ""
    score: float = 0.0
    matched_modes: list[UnifiedSearchMode]
    item: dict[str, Any]


class UnifiedSearchResponse(SQLModel):
    """返回统一排序序列、四库分组和真实计数。"""

    query: str
    selected_sources: list[UnifiedSearchSource]
    fulltext: bool
    semantic: bool
    results: list[UnifiedSearchResultOut]
    groups: dict[str, list[UnifiedSearchResultOut]]
    counts: dict[str, int]
    total: int

File: editor/src/types/unifiedSearch.ts
/**
 * Four-library unified search domain types.
 *
 * Usage:
 * Shared by the API client, workspace store, unified result rows, and split
 * native-card presentation.
 */

export const SEARCH_SOURCES = ['files', 'library', 'components', 'literature'] as const

export type SearchSource = (typeof SEARCH_SOURCES)[number]
export type SearchMatchMode = 'title' | 'fulltext' | 'semantic'

export interface UnifiedSearchResult {
  /** Stable identifier inside its source library. */
  id: string
  /** Library that owns this result. */
  source: SearchSource
  /** Display title shared by both presentation modes. */
  title: string
  /** Best lexical or semantic evidence snippet. */
  snippet: string
  /** Path, URL, or owning form location. */
  locator: string
  /** Source modification timestamp when available. */
  updated_at: string
  /** Backend relevance score after all enabled paths are merged. */
  score: number
  /** Enabled search paths that matched this same resource. */
  matched_modes: SearchMatchMode[]
  /** Original source DTO used by the split native card. */
  item: Record<string, unknown>
}

export interface UnifiedSearchResponse {
  /** Normalized query echoed by the backend. */
  query: string
  /** Sources actually searched by the backend. */
  selected_sources: SearchSource[]
  /** Whether partial full-content matching ran. */
  fulltext: boolean
  /** Whether vector retrieval and reranking ran. */
  semantic: boolean
  /** All source results in one backend-ranked sequence. */
  results: UnifiedSearchResult[]
  /** Same result objects grouped for split presentation. */
  groups: Record<SearchSource, UnifiedSearchResult[]>
  /** Real result count for every source. */
  counts: Record<SearchSource, number>
  /** Total count across selected sources. */
  total: number
}
